dotnet-test command

piotrpMSFT opened the pull request:
Fixes issue #342 
Enables CLI users and Visual Studio both to discover and execute tests given an implementation of dotnet-test-{testrunner}

dnfclas commented:
Hi __@piotrpMSFT__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

krwq commented:
Someone gave you feedback to remove an empty line? (29823fd) That's a mean feedback :)

Which files/commits should we look at? Looking at all of it is quite too much for a single PR. If you feel like it will just work I think we should just merge it and then if someone wants something different they should send a PR.

akoeplinger commented on src/Microsoft.Extensions.Testing.Abstractions/DIA/DataKind.cs:
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
```
Why Apache? Is this specific to the DIA part?

piotrpMSFT replied:
great catch. Fixed.

akoeplinger replied:
Looks like this one is still Apache :smile:

piotrpMSFT replied:
fixed too. Something threw off Ctrl+Shift+H :-1

piotrpMSFT commented:
@krwq more like a kdiff win! Fixed.

piotrpMSFT commented:
@krwq I'm mostly looking for style, naming, pattern feedback. As we get some experience with this command we'll want to augment it.

piotrpMSFT commented:
@dotnet-bot test this please

anurse commented:
Just so I'm interpreting the changes correctly: This is just about the test-runner launcher and basic design-time protocol, right? I've seen some questions about the deployment model for testing that haven't been answered and I'm hoping there will be some kind of design meeting for that.

Assuming I'm interpreting the changes right, things look fairly good. I'd like to see the protocol stuff abstracted into an assembly so that the test runner doesn't have to re-implement it.

Sridhar-MS commented on src/Microsoft.Extensions.Testing.Abstractions/DIA/DiaDataSource.cs:
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
```
@piotrpMSFT Can we upload all the DIA code as its own nuget package and reference it? This is some unnecessary bloat in our repo.

Sridhar-MS replied:
Also DIA stuff is very windows specific.

piotrpMSFT replied:
I don't want to ship an SDK on behalf of the debugger folks. Let's make it clear these are specific to our test abstractions.

anurse replied:
Yeah, these are also pretty minimal. It's a lot of files but most are just COM Interop interfaces

brthor commented on debian_config.json:
```diff
@@ -35,12 +35,13 @@
         "bin/dotnet" : "usr/bin/dotnet",
         "bin/dotnet-compile" : "usr/bin/dotnet-compile",
         "bin/dotnet-compile-csc" : "usr/bin/dotnet-compile-csc",
+        "bin/dotnet-compile-native" : "/usr/bin/dotnet-compile-native",
```
This is the wrong file, and I need to delete it (going to do now). Check packaging/debian/debian_config.json

piotrpMSFT replied:
updated the right file.

brthor commented on src/Microsoft.DotNet.Cli.Utils/AnsiConsole.cs:
```diff
@@ -1,5 +1,5 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
-// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
```
Made this change as well in the E2E test, may conflict after it merges, just FYI.

Sridhar-MS commented on src/Microsoft.DotNet.Tools.Test/Program.cs:
```diff
@@ -0,0 +1,304 @@
+            app.HelpOption("-?|-h|--help");
+
+            var parentProcessIdOption = app.Option("--parentProcessId", "Parent process ID.", CommandOptionType.SingleValue);
+            var portOption = app.Option("--port", "Port number to listen for a connection.", CommandOptionType.SingleValue);
```
Can we have a better description for `parentProcessId` and `port`? Why are these needed for running my unit tests?

Sridhar-MS replied:
If these are used only for visual studio integration, then it must be specified in the description.

piotrpMSFT replied:
fixed.

brthor commented on src/Microsoft.DotNet.Tools.Test/Program.cs:
```diff
@@ -0,0 +1,304 @@
+
+            app.HelpOption("-?|-h|--help");
+
+            var parentProcessIdOption = app.Option("--parentProcessId", "Parent process ID.", CommandOptionType.SingleValue);
```
I thought we were settling on System.CommandLine?

anurse replied:
Yeah, may as well ensure new stuff uses that.

anurse replied:
We use [`kebab-case`](https://en.wikipedia.org/wiki/Letter_case) for other command line args. We should do the same here

Sridhar-MS commented on src/Microsoft.DotNet.Tools.Test/Program.cs:
```diff
@@ -0,0 +1,304 @@
+
+            var parentProcessIdOption = app.Option("--parentProcessId", "Parent process ID.", CommandOptionType.SingleValue);
+            var portOption = app.Option("--port", "Port number to listen for a connection.", CommandOptionType.SingleValue);
+            var projectPath = app.Argument("<PROJECT>", "The project to compile, defaults to the current directory. Can be a path to a project.json or a project directory.");
```
Is this the project to 'compile' or to 'test'?

piotrpMSFT commented:
@anurse check out 
```
src\Microsoft.Extensions.Testing.Abstractions\StreamingTestDiscoverySink.cs
src\Microsoft.Extensions.Testing.Abstractions\StreamingTestExecutionSink.cs
```
These should do exactly what you're describing.

This is indeed the launcher + VS integration. At the end of the day it just launches `dotnet-test-{TestRunner}`. I'm not including a runner in this PR since I've been focusing on xunit and that runner needs to come from their repo. There will be another PR to xunit quite soon, once we're all happy with the design.

I added the ability to find commands in NuGet packages [PR #341] so we can have runners packaged as nupkg's and get picked up per-project. This is still blocked on @davidfowl's https://github.com/dotnet/cli/compare/davidfowl/loader changes since we currently have to drop the test assembly + dependencies next to the runner.

anurse commented:
:+1:

Sridhar-MS commented on src/Microsoft.DotNet.Tools.Test/Program.cs:
```diff
@@ -0,0 +1,304 @@
+                        RunConsole(projectContext, app, testRunner);
+                    }
+
+                    return 0;
```
It must return the exit code of the test runner. Returning zero makes it difficult to understand result of the test run.

anurse commented on src/Microsoft.DotNet.Tools.Test/TestHostTracing.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Microsoft.DotNet.Tools.Test
+{
+    public static class TestHostTracing
```
We should consolidate our tracing APIs. Right now we use an ad-hoc combination of things, but we have `TraceSource` and [`Microsoft.Extensions.Logging`](https://www.github.com/aspnet/Logging) to consider.

anurse commented on src/Microsoft.Extensions.Testing.Abstractions/ProjectContextExtensions.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace Microsoft.Extensions.Testing.Abstractions
+{
+    public static class ProjectContextExtensions
```
This can totally live in `Microsoft.DotNet.ProjectModel`, it's general enough IMO. Also, we can have `Get` prefix on these methods since they are methods, not properties (*sigh* extension properties :)).

anurse commented on src/Microsoft.Extensions.Testing.Abstractions/StreamingTestDiscoverySink.cs:
```diff
@@ -0,0 +1,30 @@
+            _stream = new LineDelimitedJsonStream(stream);
+        }
+
+        public void SendTest(Test test)
```
Let's name these after the message type (`TestFound`) so it's easy to match up the messages with what writes them.

anurse commented on src/Microsoft.Extensions.Testing.Abstractions/StreamingTestExecutionSink.cs:
```diff
@@ -0,0 +1,72 @@
+            _runningTests = new ConcurrentDictionary<string, TestState>();
+        }
+        
+        public void RecordStart(Test test)
```
Let's name these after the message type (`TestStarted`) so it's easy to match up the messages with what writes them.

anurse commented on src/Microsoft.Extensions.Testing.Abstractions/StreamingTestExecutionSink.cs:
```diff
@@ -0,0 +1,72 @@
+            });
+        }
+
+        public void RecordResult(TestResult testResult)
```
Let's name these after the message type (`TestResult`) so it's easy to match up the messages with what writes them.

anurse commented on src/Microsoft.Extensions.Testing.Abstractions/TestOutcome.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Microsoft.Extensions.Testing.Abstractions
+{
+    public enum TestOutcome
```
MSTest has a concept of `Inconclusive` as well. We might want to make sure we review the possible outcomes from various runners.

anurse commented:
Some fairly small comments, but otherwise looks good to me.

livarcocc commented on src/Microsoft.DotNet.Tools.Test/Program.cs:
```diff
@@ -0,0 +1,304 @@
+
+                    var testRunner = projectContext.ProjectFile.TestRunner;
+                    
+                    if (portOption.HasValue())
```
I would wrap this if/else into its own method with a good name and push the decision on what to run to there. Even better would be to push this branching into its own abstraction. But I think a separate method is good enough.

livarcocc replied:
Actually, after seen the RunDesignTime implementation I do believe we should have separate abstractions for each.

piotrpMSFT replied:
You mentioned you were looking at a cleaner model for this scenario. I'll wait on making this change until that's locked down.

livarcocc commented on src/Microsoft.Extensions.Testing.Abstractions/LineDelimitedJsonStream.cs:
```diff
@@ -0,0 +1,22 @@
+using System.IO;
```
Missing header.

piotrpMSFT replied:
fixed.

livarcocc commented on src/Microsoft.Extensions.Testing.Abstractions/Messages/Message.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        public string MessageType { get; set; }
+
+        public JToken Payload { get; set; }
```
Why not just use a string here? Why do you depend on JToken?

piotrpMSFT replied:
I'm avoiding changes to the interface between VS and dotnet-test [as it's the same as was in DNX test].

anurse replied:
Generally it reduces parsing overhead. The JSON parser has already parsed the entire message to deserialize it into the `Message` class. Storing it as a string means that when `ToObject<T>` is called (or equivalent), the string has to be re-parsed and then mapped to the `T` object. When it's a `JToken`, the parsing is still done and the data just has to be mapped.

It's a micro-optimization for sure and we could probably refine this code, but I'd agree with @piotrpMSFT that leaving the DNX code as-is for now is a good idea to ensure we aren't missing something.

livarcocc commented on src/Microsoft.Extensions.Testing.Abstractions/Messages/ProtocolVersionMessage.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace Microsoft.DotNet.Tools.Test
+{
+    public class ProtocolVersionMessage
```
This has message in the name but has no relation to what seems like a message base class above. Seems confusing.

piotrpMSFT replied:
Agreed. I'll look at cleaning up this whole VS/tool interface in a future PR.

livarcocc commented on src/Microsoft.Extensions.Testing.Abstractions/Messages/RunTestsMessage.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Microsoft.Extensions.Testing.Abstractions
+{
+    public class RunTestsMessage
```
Same here.

piotrpMSFT replied:
Agreed. I'll look at cleaning up this whole VS/tool interface in a future PR.

livarcocc commented on src/Microsoft.Extensions.Testing.Abstractions/ProjectContextExtensions.cs:
```diff
@@ -0,0 +1,31 @@
+﻿using System.IO;
```
Missing header.

piotrpMSFT commented:
Review status: 0 of 69 files reviewed at latest revision, 15 unresolved discussions.

---

*[debian_config.json, line 38 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3SdO6Lmbr8X1Ka-r1-38:-K4PRtpEMNmdRNvOCVrJ:-1124542427) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/debian_config.json#L38)):*
Fixed.

if the developer guidance has changed, can you also please update the doc? 
`Documentation/developer-guide.md`

---

*[src/Microsoft.DotNet.Tools.Test/Program.cs, line 34 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwI-r1-34:-K4PT3Skk1Tb-4Jn7C2E:27994740) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.DotNet.Tools.Test/Program.cs#L34)):*
These params are part of the arrangement with Visual Studio. We can add kebab-case in the future but must maintain compatibility.

I stayed with the DNX Cmd Line parser for the time being to minimize work that goes into the initial check-in. I'll port this to System.CommandLine soon.

---

*[src/Microsoft.DotNet.Tools.Test/Program.cs, line 35 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwI-r1-35:-K4PTpE1LsZ8htvlJNcU:-1620474753) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.DotNet.Tools.Test/Program.cs#L35)):*
fixed.

---

*[src/Microsoft.DotNet.Tools.Test/Program.cs, line 36 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwI-r1-36:-K4PTqHpYlW3muUJrWCD:-1620474753) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.DotNet.Tools.Test/Program.cs#L36)):*
fixed.

---

*[src/Microsoft.DotNet.Tools.Test/Program.cs, line 77 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwI-r1-77:-K4PVzvKTZSL4Ib2hi7B:-1802339078) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.DotNet.Tools.Test/Program.cs#L77)):*
Fixed. When running under an IDE it will return 0. When running from cmd line, it will pass through the value.

---

*[src/Microsoft.DotNet.Tools.Test/TestHostTracing.cs, line 9 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwL-r1-9:-K4PWDoDggVG1d5p_nhG:-156919078) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.DotNet.Tools.Test/TestHostTracing.cs#L9)):*
Create an issue?

---

*[src/Microsoft.Extensions.Testing.Abstractions/DIA/DiaDataSource.cs, line 1 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwO-r1-1:-K4PWRhW_zuK1a_HhIdR:-891998622) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/DIA/DiaDataSource.cs#L1)):*
do you know if the debugger team plans to make this available as a package?

---

*[src/Microsoft.Extensions.Testing.Abstractions/LineDelimitedJsonStream.cs, line 1 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwo-r1-1:-K4PW_QRH15OMdsEANoH:-1620474753) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/LineDelimitedJsonStream.cs#L1)):*
fixed.

---

*[src/Microsoft.Extensions.Testing.Abstractions/Messages/Message.cs, line 13 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwq-r1-13:-K4Pubqh5yvUIbZdyVYl:1515217022) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/Messages/Message.cs#L13)):*
I'm leaving the contract between VS and dotnet-test the same as in DNX for the time being.

---

*[src/Microsoft.Extensions.Testing.Abstractions/Messages/ProtocolVersionMessage.cs, line 6 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwr-r1-6:-K4Put1ISqI11bU3bJV6:343035917) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/Messages/ProtocolVersionMessage.cs#L6)):*
let's revisit this in a separate PR.

---

*[src/Microsoft.Extensions.Testing.Abstractions/Messages/RunTestsMessage.cs, line 8 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAws-r1-8:-K4PuxNciUkLQCDF5Qa6:343035917) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/Messages/RunTestsMessage.cs#L8)):*
let's revisit this in a separate PR.

---

*[src/Microsoft.Extensions.Testing.Abstractions/ProjectContextExtensions.cs, line 6 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwu-r1-6:-K4PXqN1eTOwSZ9rIYtH:-1397302566) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/ProjectContextExtensions.cs#L6)):*
fixed. I merged the extension methods into the ProjectContext class itself.

---

*[src/Microsoft.Extensions.Testing.Abstractions/StreamingTestDiscoverySink.cs, line 16 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwy-r1-16:-K4PYFKWU5vLbiZR9nlJ:-469593337) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/StreamingTestDiscoverySink.cs#L16)):*
done.

---

*[src/Microsoft.Extensions.Testing.Abstractions/TestOutcome.cs, line 6 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAx1-r1-6:-K4PYYzPtp0LL2WFnVHQ:-1930038374) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/TestOutcome.cs#L6)):*
Since this is a port of the existing integration into VS, let's leave it as-is for the baseline and expand as needed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dotnet/cli/344)*
<!-- Sent from Reviewable.io -->

piotrpMSFT commented:
Review status: 0 of 69 files reviewed at latest revision, 15 unresolved discussions.

---

*[debian_config.json, line 38 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3SdO6Lmbr8X1Ka-r1-38:-K4PvqDPvWLYFJ89ronw:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/debian_config.json#L38)):*
Done.

---

*[src/Microsoft.DotNet.Tools.Test/Program.cs, line 34 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwI-r1-34:-K4Pvr7e-PlFaMvv8BB7:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.DotNet.Tools.Test/Program.cs#L34)):*
Done.

---

*[src/Microsoft.DotNet.Tools.Test/Program.cs, line 35 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwI-r1-35:-K4PvrjH8zR-ycWig1P0:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.DotNet.Tools.Test/Program.cs#L35)):*
Done.

---

*[src/Microsoft.DotNet.Tools.Test/Program.cs, line 36 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwI-r1-36:-K4PvsI1GMrXYU4ioP2F:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.DotNet.Tools.Test/Program.cs#L36)):*
Done.

---

*[src/Microsoft.DotNet.Tools.Test/Program.cs, line 77 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwI-r1-77:-K4PvtWx7VLo3GvTtZgO:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.DotNet.Tools.Test/Program.cs#L77)):*
Done.

---

*[src/Microsoft.Extensions.Testing.Abstractions/LineDelimitedJsonStream.cs, line 1 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwo-r1-1:-K4PvydHKsaL6eVLsa0V:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/LineDelimitedJsonStream.cs#L1)):*
Done.

---

*[src/Microsoft.Extensions.Testing.Abstractions/Messages/Message.cs, line 13 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwq-r1-13:-K4Pw-3gm_ujFWYenZfc:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/Messages/Message.cs#L13)):*
Done.

---

*[src/Microsoft.Extensions.Testing.Abstractions/Messages/ProtocolVersionMessage.cs, line 6 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwr-r1-6:-K4Pw-zb_XHyyUscGivb:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/Messages/ProtocolVersionMessage.cs#L6)):*
Done.

---

*[src/Microsoft.Extensions.Testing.Abstractions/Messages/RunTestsMessage.cs, line 8 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAws-r1-8:-K4Pw0aovN7V3GAFmrBD:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/Messages/RunTestsMessage.cs#L8)):*
Done.

---

*[src/Microsoft.Extensions.Testing.Abstractions/ProjectContextExtensions.cs, line 6 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwu-r1-6:-K4Pw1aIXd_BXJJ5e9iE:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/ProjectContextExtensions.cs#L6)):*
Done.

---

*[src/Microsoft.Extensions.Testing.Abstractions/StreamingTestDiscoverySink.cs, line 16 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwy-r1-16:-K4Pw2JGlytWsGiQt2Ut:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/StreamingTestDiscoverySink.cs#L16)):*
Done.

---

*[src/Microsoft.Extensions.Testing.Abstractions/TestOutcome.cs, line 6 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAx1-r1-6:-K4Pw3Nmw126UIcFLvIw:-499146009) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.Extensions.Testing.Abstractions/TestOutcome.cs#L6)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dotnet/cli/344)*
<!-- Sent from Reviewable.io -->

anurse commented on src/Microsoft.DotNet.Tools.Test/Program.cs:
```diff
@@ -0,0 +1,305 @@
+
+            app.HelpOption("-?|-h|--help");
+
+            var parentProcessIdOption = app.Option("--parentProcessId", "Used by IDEs to specify their process ID. Test will exit if the parent process does.", CommandOptionType.SingleValue);
```
reviving my earlier `kebab-case` comment ;). Can we make it `--parent-process-id` for consistency?

piotrpMSFT replied:
[sorry, I had commented on all of this but Reviewable didn't do what I thought it'd do :-/]

The issue is that these are the commands issued by VS when calling DNX test and therefore DotNet-test. Let's make sure we get IDE integration working and adjust in lock-step.

anurse replied:
Ah, fair enough. If there is already a "protocol" for these arguments in place, that makes sense. Can you put a small comment in the code to that effect to prevent "helpful" people from breaking it in the future?

Yeah... we tried Reviewable once on ASP.NET, but found it hard to get things in sync and most people don't bother clicking the "review on Reviewable" button :).

piotrpMSFT commented:
*[src/Microsoft.DotNet.Tools.Test/Program.cs, line 61 \[r1\]](https://reviewable.io:443/reviews/dotnet/cli/344#-K4PRk3TpoVBFvATaAwI-r1-61:-K4Q-76cmJau5ws7nJNi:-1487732582) ([raw file](https://github.com/dotnet/cli/blob/fb68e5191b1910b11949c8ceb74bf8ef518fb2ff/src/Microsoft.DotNet.Tools.Test/Program.cs#L61)):*
You mentioned you were looking at  a cleaner model for this scenario. I'll wait on making this change until that's locked down.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dotnet/cli/344)*
<!-- Sent from Reviewable.io -->

piotrpMSFT commented:
@dotnet-bot test this please

anurse commented:
/cc @victorhurdugaci
